### PR TITLE
F/short subaccount id change

### PR DESCRIPTION
**Pull request by Tofel**
Convert subaccount_id to decimal when **serializing** short subaccount id.

### CODE REVIEW

File: packages/injective-cosmwasm/src/exchange/types.rs [hunk 1]
```diff
@@ -258,6 +272,22 @@ impl<'de> Deserialize<'de> for ShortSubaccountId {
     }
 }
 
+impl Serialize for ShortSubaccountId {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        // let's check first check again if it's a hexadecimal number and parse to decimal (expected by the chain)
```
Comment (gorgos) [inline packages/injective-cosmwasm/src/exchange/types.rs]:
```suggestion
        // chain expects decimal number, but ShortSubaccountId stores hexadecimal string
```

File: packages/injective-cosmwasm/src/exchange/types.rs [hunk 2]
```diff
@@ -258,6 +272,22 @@ impl<'de> Deserialize<'de> for ShortSubaccountId {
...
+        let as_decimal = match u32::from_str_radix(self.0.as_str(), 16) {
+            Ok(dec) => Ok(dec),
+            Err(_) => Err(S::Error::custom("Invalid value: ShortSubaccountId was not a hexadecimal number")),
```
Comment (gorgos) [inline packages/injective-cosmwasm/src/exchange/types.rs]: why not add the `self.0` to the error?

File: packages/injective-cosmwasm/src/exchange/types.rs [hunk 3]
```diff
@@ -258,6 +272,22 @@ impl<'de> Deserialize<'de> for ShortSubaccountId {
...
+        };
```
Comment (gorgos) [inline packages/injective-cosmwasm/src/exchange/types.rs:300]: I would add the `?` here.

File: packages/injective-cosmwasm/src/exchange/types.rs [hunk 4]
```diff
@@ -258,6 +272,22 @@ impl<'de> Deserialize<'de> for ShortSubaccountId {
...
+
+        // if the string cannot be parsed to a decimal return error otherwise let's pad it with left zeros
```
Comment (gorgos) [inline packages/injective-cosmwasm/src/exchange/types.rs]: unnecessary comment

File: packages/injective-cosmwasm/src/exchange/types.rs [hunk 5]
```diff
@@ -218,7 +219,20 @@ impl ShortSubaccountId {
...
+        };
+
+        // if the string cannot be parsed to a decimal return error
+        let decimal_u32 = as_decimal?;
+
+        // if it is a hexadecimal number, let's pad it with left zeros
+        let number_to_use = format!("{:0>3}", decimal_u32.to_string());
+
+        // check if resulting number isn't bigger than chain limit
+        match number_to_use.parse::<u16>() {
```
Comment (gorgos) [inline packages/injective-cosmwasm/src/exchange/types.rs]: I don't think we need this, do we? if we already do the conversion to decimal during serializations.

- If we do, please say why and remove the comments.
- If we don't, please remove here.

**Review by gorgos — CHANGES_REQUESTED** [review]

File: packages/injective-cosmwasm/src/exchange/types.rs (hunk 5 above)
Comment (Tofel) [reply packages/injective-cosmwasm/src/exchange/types.rs]: Well, I think we should validate the input, when a new instance is created. You don't want to find out that input is invalid only when you're serializing it, which would probably when submessage is dispatched? 

This part could deffo be removed here:
```
       // if it is a hexadecimal number, let's pad it with left zeros
        let number_to_use = format!("{:0>3}", decimal_u32.to_string());
```

Also, I will make that code more concise. Plus, add a method `must_new()` which will panic on invalid input. I that once this PR is merged we should change all usages of `unchecked()` for `must_new()`: https://github.com/InjectiveLabs/cw-injective/issues/160

**Comment by Tofel** [comment]
@gorgos added support for deserializing from long subaccount id, added some tests and added validate function (that's now used by `new()`, `must_new()` and `serialize()` and `deserialize()`.

**Review by gorgos — CHANGES_REQUESTED** [review]
The changes LGTM, but I thought we discussed also adding a new feature to deserialize a long subaccount id into the short type?

**Review by gorgos — APPROVED** [review]

**Pull request merged**